generate_analysis_graphs.py

```diff
@@ -172,6 +172,42 @@
     # TODO: Update the graph generation code to use the engagement_counts dict rather than performing additional local
     #       derivations of the participation figures
 
+    log.info("Computing the participation frequencies...")
+    repeat_participations = OrderedDict()
+    for i in range(1, len(PipelineConfiguration.RQA_CODING_PLANS) + 1):
+        repeat_participations[i] = {
+            "Episodes Participated In": i,
+            "Number of Individuals": 0,
+            "% of Individuals": None
+        }
+
+    # Compute the number of individuals who participated each possible number of times, from 1 to <number of RQAs>
+    # An individual is considered to have participated if they sent a message and didn't opt-out, regardless of the
+    # relevance of any of their messages.
+    for ind in individuals:
+        if ind["consent_withdrawn"] == Codes.FALSE:
+            weeks_participated = 0
+            for plan in PipelineConfiguration.RQA_CODING_PLANS:
+                if plan.raw_field in ind:
+                    weeks_participated += 1
+            assert weeks_participated != 0, f"Found individual '{ind['uid']}' with no participation in any week"
+            repeat_participations[weeks_participated]["Number of Individuals"] += 1
```

Comment:
I don't remember if python recommends canonicalising int keys to dictionaries as strings or if that is only for JSON

Reply:
JSON requires keys to be strings, Python allows any type that is hashable. 

I think using ints is the right thing to do here - it's the sensible way to think about the data and the dictionary is short-lived (i.e. only used over a few lines of code) so there isn't much scope for confusion.

I will leave as-is for now but happy to change if you think I should.